src/sb-util.ts

```diff
@@ -1,32 +1,181 @@
-import { Queryable, ScratchProjectOptions, AssetFetcher } from './abstracts';
-import { ProjectSource, Sb3Fetcher, ProjectJsonFetcher, ProjectByCloudIdFetcher } from './asset-fetcher';
+import { Queryable, AssetFetcher, SpriteProperties, SpritePosition, BlockProperties } from './abstracts';
+import { Sb3Fetcher, ProjectJsonFetcher, ProjectByCloudIdFetcher } from './asset-fetcher';
+import { validateSpriteSelector, isSelectorAttrValue, getAttrValue, attrValueContainsQuotes } from './selector-parse';
+import { map, filter, makeIterable, first } from './generators';
 
-export class ScratchProject implements Queryable {
+enum CollectionTypes {
+	SPRITES = 'sprites',
+	BLOCKS = 'blocks',
+	ASSETS = 'assets'
+}
+
+enum ScratchProjectKeys {
+	TARGETS = 'targets'
+}
+
+enum SpriteAttributes {
+	BLOCKS = 'blocks',
+	BROADCASTS = 'broadcasts',
+	LISTS = 'lists'
+}
+
+enum BlockAttributes {
+	ID = 'id'
+}
+
+/*
+sb-util CLASSES.
+
+The classes below make up the sb-util API.
+*/
+
+// This is the global storage for project data
+const storage = new WeakMap();
+
+export class ScratchProject {
 	constructor(projectJSON: Object, assetFetcher?: AssetFetcher) {
+		storage.set(this, projectJSON);
+	}
+
+	prop(property: string) {
+		return storage.get(this)[property];
+	}
+
+	sprites(selector?: string){
+		if (selector && typeof selector !== 'string') throw new Error('SpriteCollection selector should be a string!');
+
+		const sprites = this.prop(ScratchProjectKeys.TARGETS);
+		if (!selector && typeof selector !== 'string') return new SpriteCollection(sprites);
+		return new SpriteCollection(sprites).query(selector);
+	}
+
+	stage() {
+		return this.sprites('[isStage=true]').first();
+	}
+
+	blocks() {
+		throw new Error('ScratchProject blocks() function not implemented yet!');
+	}
+}
+
+export class SpriteCollection implements Queryable  {
+	constructor(sprites: Iterable<SpriteProperties>){
+		storage.set(this, sprites);
+	}
+
+	first(): Sprite {
+		const props: SpriteProperties = first(storage.get(this));
+		return props ? new Sprite(props) : null;
+	}
+
+	prop(attribute: string) {
+		const first = storage.get(this).slice().shift();
+		return first[attribute];
 	}
 
-	query(selector?: String) {
-		return [];
+	/*
+		Currently, the query selector syntax only supports
+		attribute selector style in the form of [attr] and
+		[attr=value]
+	*/
+	query(selector: string){
+		validateSpriteSelector(selector);
+
+		// string between brackets
+		const selectorBody = selector.substring(1, selector.length-1);
+
+		let sprites, filterFunction, allSprites = storage.get(this);
+
+		// case when selector string is in [attr=value] form
+		if(isSelectorAttrValue(selectorBody)) {
+			const [attr, valueString] = getAttrValue(selectorBody);
+
+			this['value'] = valueString;
```

Comment:
this.value brings about a syntax error, so i put this in an variable called attrValue

Reply:
I'm guessing this.value was a typescript syntax error as opposed to a javascript syntax error. Because the constructor does not define a value member on SpriteCollection, it is a typescript syntax error to assign this.value. `this['value']` probably doesn't have a type check implemented yet, or is intentionally left unchecked.

So we should avoid `this['someMember']` uses as we'll lose typescript checks. Also value shouldn't be assigned to this but assigned to the new SpriteCollection if you were to store it on a collection. Storing on the collection calling query would create a race condition if you made two queries on the same collection.